### DISCUSSION TITLE
Native Image Committer Community Meeting 2021-09-20

### DISCUSSION THREAD

**Discussion by christianwimmer** [oracle/graal#3818] [Show and tell]
List of all past and upcoming meetings: #3933

## New and Noteworthy
* JDK 8 support is going to be deleted. After the 21.3 release, we will gradually remove all testing on JDK 8 and then delete all the JDK 8 specific code before the 22.0 release. No PR yet though.
* Configuration file entries that are necessary at image build time for reflection, JNI, resources, and Proxy, can now be conditional based on the reachability of class. This allows a more precise configuration, which can reduce the size of a native image. [Documentation](https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/Reflection.md#conditional-configuration) with more details and an example for a conditional reflection configuration. 
  #3606
  #3774
  #3803
  
* A native image now always contains reflection metadata for all methods that were seen as reachable by the static analysis. Calling `Class.getDeclaredMethods` (and all the other variants to look up methods and constructors) now returns all reachable methods for all classes. Previously, only methods that were explicitly registered for reflective invocation were returned. This new feature allows significant reduction of the manual reflection configuration: only methods that need to be invoked via `Method.invoke`, or methods that would not be seen as reachable based on reachability, need to be registered manually.
   #3637

* Added several new optimizations to reduce the image size as well as the image build time for large applications.
  #3765
  #3777
  #3529
  #3771
  #3595
  #3641

* Initial support for the Java Platform Module System: The native image generator now accepts the module-related arguments known from the `java` launcher like `-m`, `-p`, and `--add-opens`. When such a module-related argument is used, then the image generator itself is used as a module too. There were many PRs for this over the last months. Here are just a few of them:
  #3763
  #3622
  #3445
  cf56a336615fe9d6f403e26b7b250c79256e3297
  #3552
  #3446
  #3382

* Added a new policy for the [Serial GC](https://www.graalvm.org/reference-manual/native-image/MemoryManagement/#serial-garbage-collector) reducing the memory footprint of applications at run time. The new policy enables survivor spaces for the young generation. In addition, a full GC no longer scans the whole image heap for root pointers, but only parts of the image heap that have been written to.
  #3589
  f3c2685ed1e19e14bf693b3d43c8c50443b559db

* Small methods are now inlined before the static analysis to improve the precision of the static analysis.
  ab9bfcf776020dd9bf73673122bec6dd8898aabd
  #3640 
  #3628
  #3545
 
* Arguments to the native image generator can now be passed in a separate file that is provided with the `@argument` option. This is a necessary workaround for platforms like Windows where the the length of the command line is limited.
  #3734
* Root certificates can now be configured at image run time.  [Documentation](https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/CertificateManagement.md#runtime-options) with details and a list of system properties that are now read at run time to configure the root certificates.
  #3652
  #1999
* Crash dumps, which are printed when a segfault or a fatal VM error happen, now include more details to aid diagnostics.
  #3515
  #3795
  #3741
* The beginning of the image heap is now protected memory so that null pointer accesses produce a segfault instead of accessing random memory.
  e4120344ec9724de926e7349e3a9f86e627eeaea
* Internal VM errors, where an illegal virtual method table entry is accessed, produce a better crash dump.
  #3702
* On Linux, the processor count reported at run time now takes limitations set via `cpuset` into account.
  #3762

* Improved Dwarf debug info generation now includes information about inlined methods.
  #3745
* JDK Flight Recorder (JFR) support is now also available for JDK 17.
  #3726

### Parse bytecode only once

Historically, the bytecode of all methods was parsed twice during an image build: for static analysis, and then again for AOT compilation. That had advantages for images with Truffle languages that need deoptimization and JIT compilation (and is done that way for such images today). But over time, we ran into more and more subtle problems: the two parsings can get out of sync. Then code that the static analysis did not see as reachable suddenly got AOT compiled, leading to strange errors at run time.

aa2f6e89ce28a4f84239bf6ffb83941fb8737a3a is the switch to "parse once".
* Static analysis results are now applied to the Graal graphs that were parsed for analysis: [StrengthenGraphs](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/results/StrengthenGraphs.java)
* Graal graphs are "transplanted" from the "analysis universe" into the "hosted universe": [CompileQueue.transplantGraph](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/CompileQueue.java#L744)

### Inline methods before static analysis

One of the main reasons for "parse once" was to allow method inlining before static analysis. We experimented with inlining before analysis beforehand, but the first implementation (before "parse once") was not reliable and is now removed (#3545).
The new inlining before analysis is a multi-step process:
* All graphs are parsed without method inlining first: [AnalysisParsedGraph](https://github.com/oracle/graal/blob/7d97b4884726133a39b0888dbaa4711a544d432f/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/flow/AnalysisParsedGraph.java#L88)
* The parsed graph is immediately encoded into a more compact representation, also used by the Truffle partial evaluator.
* When the [MethodTypeFlowBuilder](https://github.com/oracle/graal/blob/7d97b4884726133a39b0888dbaa4711a544d432f/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/flow/MethodTypeFlowBuilder.java#L171) processes a method that is reachable, the encoded graph is decoded again, and method inlining is done during decoding: [InlineBeforeAnalysis](https://github.com/oracle/graal/blob/7d97b4884726133a39b0888dbaa4711a544d432f/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/phases/InlineBeforeAnalysis.java#L118)

The policy for inlining before analysis is quite conservative: [InlineBeforeAnalysisPolicyImpl](https://github.com/oracle/graal/blob/a9299bbed927be49ae18fd9f506dd4d5e558c654/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/InlineBeforeAnalysisPolicyImpl.java) only inlines methods with one computational Graal node. This can be a method invocation (so "forwarding" method calls are inlined), a field access, arithmetic, ... Constants do not count to the limit. We can never reduce inlining because that would be a backward-incompatible change!

Note that only methods that can be statically bound without any static analysis or speculation can be inlined before the static analysis. So this is limited to `static` methods, or `final` / `private` instance methods. But it is enough for some important use cases in Spring where reflective method lookups needs to be constant folded: https://github.com/oracle/graal/issues/2500#issuecomment-860330959

### Image heap scanning

Currently, the static analysis scans the image heap repeatedly: [PointsToAnalysis.finish](https://github.com/oracle/graal/blob/37e18768edcd82552607b6504f522cdbd13f5f65/substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/PointsToAnalysis.java#L644). This is time consuming for no good reason (other than "we have always done it that way"). Image heap writing after static analysis then scans all objects again: [NativeImageHeap.addObjectToImageHeap](https://github.com/oracle/graal/blob/2ed3a6b6f5f18819309336f5ce4f28bb13ba4ae1/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageHeap.java#L365). This can lead to inconsistencies and [hard-to-debug error messages](https://github.com/oracle/graal/blob/2ed3a6b6f5f18819309336f5ce4f28bb13ba4ae1/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageHeap.java#L497).

We are in the process of changing this, and visit each object only once during static analysis to build a "shadow heap". Some pre-work for this is already merged, like #3641,

## Possible deep dive topics for next meeting
Please send suggestions, or "upvote" a suggestion, by adding a comment to this discussion.
* Reflection data for all methods
* Conditional reflection configuration
* GC policy and GC performance
* Class initialization at image build time vs. run time
* Incomplete class paths and the future of `--allow-incomplete-classpath` (there is some discussion about that in #2762)

**Comment by ziyilin** [comment]
The topics are very interesting. I'd vote for the "GC" topic for the next meeting. 
When I read the source code, I may get lost in the details. Introducing some of the big pictures such as the reasons behind the design would be very helpful.

**Comment by christianwimmer** [comment]
The next meeting will be on **Thursday, October 21**, at [17:00 Central Europe / 11:00 Eastern / 8:00 Pacific Time](https://www.timeanddate.com/worldclock/fixedtime.html?iso=20211021T08&p1=224)

**Comment by galderz** [comment]
> Incomplete class paths and the future of --allow-incomplete-classpath (there is some discussion about that in Remove some limitations of GraalVM Native Image #2762)

^ I've just created https://github.com/oracle/graal/issues/3929 with ideas to improve the usability of this configuration option.
